src/main/golem/util/validation/dimensions.kt

```diff
@@ -0,0 +1,256 @@
+            for (x in 1..nlpad)
+                append(lpad)
+        append(text)
+        if (rpad != null)
+            for (x in 1..nrpad)
+                append(rpad)
+        if (gap != null)
+            append(gap)
+    }
+
+    // Build an english-description of numeric dimensions
+    private fun concoctFixedDescription(rows: String, cols: String): String {
+        if (rows == "1")
+            return "a ${cols}-vector"
+        if (cols == "1")
+            return "a ${rows}-vector"
+        return "${rows} by ${cols}"
+    }
+
+    // Build a list of things compareTo needs to be the same as. Looks up other things that have
+    // the same character-based dimension defined.
+    private fun concoctComparison(name: String, compareTo: String): List<String> {
+        val out = mutableListOf<String>()
+        if (compareTo !in seenInError) {
+            for ((name2, dims) in declared) {
+                val (rows, cols) = dims
+                if (name2 != name) {
+                    if (rows == compareTo) out.add("as ${name2} has rows")
+                    if (cols == compareTo) out.add("as ${name2} has columns")
+                }
+            }
+            seenInError.add(compareTo)
+        }
+        return out
+    }
+
+    // Build our pretty table of dimensions real vs actual.
+    private fun concoctExceptionMessage(context: ValidationContext, failed: List<Pair<String,
+    Matrix<Double>>>): String {
+        val strings = mutableListOf<String>()
+        for ((index, matrix) in context.matrices.withIndex()) {
+            val name = context.matrixNames[index]
+            val decl = declared[name]
+            if (decl != null) {
+                strings.add(name)
+                val (r, c) = decl
+                strings.add("${r}x${c}")
+                strings.add("${matrix.numRows()}x${matrix.numCols()}")
+            }
+        }
+        val headers = arrayOf("Matrix", "Required", "Actual")
+        val widths = mutableListOf(headers[0].length, headers[1].length, headers[2].length)
+        for ((index, str) in strings.withIndex())
+            if (str.length > widths[index % widths.size])
+                widths[index % widths.size] = str.length
+        val bstream = ByteArrayOutputStream()
+        val pstream = PrintStream(bstream)
+        pstream.append("Invalid matrix dimensions.\n\n")
+        for ((index, header) in headers.withIndex())
+            pstream.writeColumn(widths[index], header, ' ', ' ')
+        pstream.append("\n")
+        for (width in widths)
+            pstream.writeColumn(width, "", '=', null)
+        for ((index, str) in strings.withIndex()) {
+            val windex = index % widths.size
+            if (windex == 0) pstream.append("\n")
+            pstream.writeColumn(widths[windex], str, if (windex == 0) null else ' ')
+        }
+        pstream.append("\n")
+        for ((name, matrix) in failed) {
+            val decl = declared[name]
+            if (decl != null) {
```

Comment:
You can replace a lot of these null checks with more idiomatic code (e.g. let, see above).